### DISCUSSION TITLE
"Failed to authenticate" for discord oauth2

### DISCUSSION THREAD

**Discussion by ulughann** [pocketbase/pocketbase#1739] [Q&A] [answered]
code related to:

js for initializing the links:

`
const pb = new PocketBase('http://127.0.0.1:8090');
const redirectUrl = 'http://127.0.0.1:3000/redirect';

async function loadLinks() {
    const authMethods = await pb.collection('users').listAuthMethods();
    const listItems = [];

    for (const provider of authMethods.authProviders) {
        const $li = $('<li><a class="button is-bordered"><img src="https://icongr.am/material/${provider.name.toLowerCase()}.svg?size=26&color=fa1e4e" width="16" height="16"><span>${provider.name} ile giriş yap</span></a></li>');

        $li.find('a')
            .attr('href', provider.authUrl + redirectUrl)
            .data('provider', provider)
            .click(function () {
                localStorage.setItem('provider', JSON.stringify($(this).data('provider')));
            });

        listItems.push($li);
    }

    $('#list').html(listItems.length ? listItems : '<li>No OAuth2 providers.</li>');
}

loadLinks();
`

html for redirect

`
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <title>OAuth2 redirect page</title>
</head>
<body>
    <pre id="content">Authenticating...</pre>

    <script src="https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@master/dist/pocketbase.umd.js"></script>
    <script type="text/javascript">
        const pb = new PocketBase("http://127.0.0.1:8090");
        const redirectUrl = 'http://127.0.0.1:3000/redirect';
        const params = (new URL(window.location)).searchParams;
        const provider = JSON.parse(localStorage.getItem('provider'))
        if (provider.state !== params.get('state')) {
            throw "State parameters don't match.";
        }

        pb.collection('users').authWithOAuth2(
            provider.name,
            params.get('code'),
            provider.codeVerifier,
            redirectUrl,
            {
                emailVisibility: false,
            }
        ).then((authData) => {
            document.getElementById('content').innerText = JSON.stringify(authData, null, 2);
        }).catch((err) => {
            document.getElementById('content').innerText = "Failed to exchange code.\n" + err;
        });
    </script>
    
</body>
</html>
`

routing in express (node with tsc)
`
app.get('/', (req: any, res: any) => {
    res.render('home');
});

app.get('/redirect', (req: any, res: any) => {
    res.render('partials/redirect');

});
`

I cannot oauth and I get the error below
![image](https://user-images.githubusercontent.com/110777803/216093930-ed99cb64-2df7-42a0-8ffe-2c1a0807ae48.png)

edit: for the button construction, I changed \` with ' to avoid markdown breaking. It has \` on the actual code

**Comment by ganigeorgiev** [comment, marked as answer]
There are several reasons the OAuth2 sign-up could fail:

- The Create API rule of your auth collection doesn't allow creating new users.
- "Always require email" collection option is enabled (could be changed from the collection settings > Options tab).

You can also start PocketBase with `--debug` for more detailed error messages and/or check the logs in the Admin UI > Logs.
